File: 백준/Silver/31246. 모바일 광고 입찰/모바일 광고 입찰.py
n, k = map(int, input().split())
infos = [list(map(int, input().split())) for _ in range(n)]

def solution(n, k, infos):
    # info = (Ai, Bi)
    # Ai = 제시한 입찰 가격 / Bi = 제시한 입찰가 외 최고 입찰가
    # Ai > Bi 입찰!!!
    arr = [0]
    cnt = 0
    for info in infos:
        A_i, B_i = info
        if A_i >= B_i:
            cnt += 1
        else:
            # 입찰가의 차이 기록
            arr.append(B_i - A_i)
    # 최소로 올리기 위함.
    arr.sort()
    # '더' 필요한 낙찰의 수
    final = k - cnt
    if final <= 0:
        return 0
    return arr[final]

print(solution(n, k, infos))

File: 백준/Silver/31246. 모바일 광고 입찰/모바일 광고 입찰.java


import java.io.*;
import java.util.*;

class Solution {
    public int solution(int N, int K, long[][] parts) {
        int answer = 0;
        int cnt = 0;

        List<Long> list = new ArrayList<>();

        for (long[] part : parts) {
            long A = part[0];
            long B = part[1];
            if (A >= B) {
                cnt++;
            } else {
                list.add(B - A);
            }
        }

        Collections.sort(list);
        
        int num = K - cnt;
        
        if (num <= 0) {
            return 0;
        }
        long x = list.get(num - 1);
        return (int) x;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        long[][] parts = new long[N][2];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            parts[i][0] = Long.parseLong(st.nextToken());
            parts[i][1] = Long.parseLong(st.nextToken());
        }

        int answer = new Solution().solution(N, K, parts);
        System.out.println(answer);
    }
}

